kernel/sched.c

```diff
@@ -1280,12 +1280,21 @@ static inline void z_vrfy_k_thread_priority_set(k_tid_t thread, int prio)
 void z_impl_k_thread_deadline_set(k_tid_t tid, int deadline)
 {
 	struct k_thread *thread = tid;
+	int32_t newdl = k_cycle_get_32() + deadline;
 
+	/* The prio_deadline field changes the sorting order, so can't
+	 * change it while the thread is in the run queue (dlists
+	 * actually are benign as long as we requeue it before we
+	 * release the lock, but an rbtree will blow up if we break
+	 * sorting!)
+	 */
 	K_SPINLOCK(&_sched_spinlock) {
-		thread->base.prio_deadline = k_cycle_get_32() + deadline;
 		if (z_is_thread_queued(thread)) {
 			dequeue_thread(thread);
+			thread->base.prio_deadline = newdl;
 			queue_thread(thread);
+		} else {
+			thread->base.prio_deadline = newdl;
```

Comment:
If the thread whose deadline we are changing is currently executing, does that not mean that we should also perform a reschedule if it's deadline was pushed farther out into the future?

Reply:
Maybe?  That's a policy question of whether or not deadlines should be a scheduling point.   The code as it has been so far says "no", presumably (I mean this is all ancient and I don't remember the design decisions) because in EDF you have to know ahead of time what your deadlines are and whether or not they can be handled in time, which isn't really consistent with "change it while it's running".  But I could totally see an argument made in the other direction too...

FWIW: in SOF the only place this gets used is in Marcin's DP scheduler code, and even then only if you have more than one component scheduled on the same core, which no currently-envisioned topologies will be doing any time soon.   Even then DP components currently have 10ms (~100 ticks!) buffer sizes and are **really** not going to see a lot of issues with timer-scale scheduling jitter.

Reply:
Sorry, but what is SOF?
I saw the word many times and still unknown the meaning.

Reply:
https://github.com/thesofproject/sof